File: leoz-central/src/main/java/org/deku/leoz/central/messaging/MessageListenerConfiguration.java
package org.deku.leoz.central.messaging;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.deku.leoz.central.data.repositories.NodeRepository;
import org.deku.leoz.messaging.activemq.ActiveMQContext;
import org.deku.leoz.node.messaging.auth.v1.IdentityMessage;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Lazy;
import sx.jms.embedded.Broker;
import sx.jms.embedded.activemq.ActiveMQBroker;

import javax.annotation.PostConstruct;
import javax.annotation.PreDestroy;
import javax.inject.Inject;

/**
 * Created by masc on 20.06.15.
 */
@Configuration
@Lazy(false)
public class MessageListenerConfiguration {
    private Log mLog = LogFactory.getLog(this.getClass());

    @Inject
    private NodeRepository mNodeRepository;

    /** Central message listener */
    MessageListener mMessageListener;

    /**
     * Broker event listener
     */
    Broker.EventListener mBrokerEventListener = new Broker.EventListener() {
        @Override
        public void onStart() {
            startIfReady();
        }

        @Override
        public void onStop() {
            stop();
        }
    };

    /**
     * Start message listener
     */
    private void startIfReady() {
        this.stop();

        if (ActiveMQContext.instance().getBroker().isStarted()) {
            // Configure and create listener
            mMessageListener = new MessageListener(ActiveMQContext.instance());

            // Add message handler delegatess
            mMessageListener.addDelegate(IdentityMessage.class,
                    new IdentityMessageHandler(
                            mNodeRepository));

            mMessageListener.start();
        }
    }

    /**
     * Stop message listener
     */
    private void stop() {
        if (mMessageListener != null) {
            mMessageListener.dispose();
            mMessageListener = null;
        }
    }

    @PostConstruct
    public void onInitialize() {
        mLog.info("Initializing node message listener");

        // Register event listeners
        ActiveMQBroker.instance().getDelegate().add(mBrokerEventListener);

        this.startIfReady();
    }

    @PreDestroy
    public void onDestroy() {
        mMessageListener.stop();
    }
}

File: leoz-central/src/main/java/org/deku/leoz/central/config/MessageListenerConfiguration.kt
package org.deku.leoz.central.config

import org.deku.leoz.config.ActiveMQConfiguration
import org.deku.leoz.config.JmsChannels
import org.slf4j.LoggerFactory
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Lazy
import sx.mq.MqBroker
import sx.mq.jms.activemq.ActiveMQBroker
import sx.mq.jms.listeners.SpringJmsListener
import java.util.concurrent.ExecutorService
import javax.annotation.PostConstruct
import javax.annotation.PreDestroy
import javax.inject.Inject

/**
 * Leoz-central message listener configuration
 * Created by masc on 20.06.15.
 */
@Configuration
@Lazy(false)
open class MessageListenerConfiguration : org.deku.leoz.node.config.MessageListenerConfiguration() {
    private val log = LoggerFactory.getLogger(this.javaClass)

    @Inject
    private lateinit var executorService: ExecutorService

    @Inject
    private lateinit var mqConfiguration: ActiveMQConfiguration

    @Inject
    private lateinit var broker: ActiveMQBroker

    /**
     * Central queue listener
     */
    val centralQueueListener by lazy {
        SpringJmsListener(
                channel = JmsChannels.central.mainQueue,
                executor = this.executorService)
    }

    /**
     * Central log queue listener
     */
    val centralLogQueueListener by lazy {
        SpringJmsListener(
                channel = JmsChannels.central.logQueue,
                executor = this.executorService)
    }

    //region Lifecycle
    /**
     * Broker event listener
     */
    private val brokerEventListener: MqBroker.EventListener = object : MqBroker.DefaultEventListener() {
        override fun onStart() {
            startIfReady()
        }

        override fun onStop() {
            stop()
        }
    }

    /**
     * Start message listener
     */
    private fun startIfReady() {
        this.stop()

        if (this.broker.isStarted) {
            this.centralQueueListener.start()
            this.centralLogQueueListener.start()
        }
    }

    /**
     * Stop message listener
     */
    private fun stop() {
        this.centralQueueListener.stop()
        this.centralLogQueueListener.stop()
    }

    @PostConstruct
    override fun onInitialize() {
        // Hook up with broker events
        this.broker.delegate.add(brokerEventListener)
        this.startIfReady()
    }

    @PreDestroy
    override fun onDestroy() {
        this.stop()
    }
    //endregion
}
